src/engine/renderer/tr_main.cpp

```diff
@@ -1871,6 +1874,10 @@ bool R_MirrorViewBySurface(drawSurf_t *drawSurf)
 	// trivially reject portal/mirror
 	if (SurfIsOffscreen(drawSurf, surfRect))
 	{
+		if ( drawSurf->shader->portalOutOfRange ) {
+			R_AddPreparePortalCmd( drawSurf );
```

Comment:
I don't understand, why do we need to do anything at all if the portal surface is completely off-screen?

Reply:
I suppose the name of the function becomes misleading in this case, since one of the cases it can return true is if the portal surface is on screen *and* it's further away than the maximum range for this portal.

Reply:
Could use a comment, or better a change to that function's name.

Reply:
Yes, the code in `SurfIsOffscreen` is returning early when the portal is out of range, before we implemented alphagen portal, doing this made the portal go black (assuming the screen was clearest at frame begin) when being out of range. Being out of range or off screen was exactly the same for a naive portal implementation without blending and without out-of-range fallback: the portal should stop being rendered. That's why `SurfIsOffscreen` says the portal is off screen when out of range (even if on screen).

But now that alphagen portal is implemented, what we do is we render the portal until it's out of range (as before), blending it with a texture while taking distance (new!). And then, once the portal is out of range, the portal stops being rendered (as before) but the texture is rendered instead without blending (new!).

Right now with the current engine design keeping that already existing `SurfIsOffScreen` mechanism to know the portal is out of range looks to be the simplest way to do it, but yes the function name is confusing.

To do things better we would need to get two different informations:

- portal is out of range
- surface is off screen

Right now `SurfIsOffscreen` gives information that “surface is off screen or on screen but out of range” information, doing otherwise may require some deep refactor that better be done later.

Reply:
If that function only did one thing, then we'd have to loop over all the vertices twice. We can also use the mechanism already established in https://github.com/DaemonEngine/Daemon/blob/7920c1a69f2aed340543d3653fd9eed24c474f91/src/engine/renderer/Material.cpp#L1865 to check if the bounding sphere is off-screen which is less precise but probably fine. It'd also fix the issue with `r_smp` (might need a slight change to how portal bounding spheres are generated in that case though). Also might need a tweak to support moving portals, but since we don't really have maps that use that we can skip that for now.

Reply:
I've changed the name to `PortalOffScreenOrOutOfRange()` now. Also added a comment explaining the need to do `R_AddPreparePortalCmd()` and `R_FinalisePortalCmd()`.